File: src/iai_mcp/lillibrain/io.py
"""Durable file-sync abstractions for the LilliBrain storage engine.

Wraps OS-level fsync primitives so callers never call os.fsync() directly.
F_FULLFSYNC on macOS forces the drive controller's DRAM cache to NAND.
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

# fcntl exists only on POSIX; it is used solely for the Darwin-only
# F_FULLFSYNC path, so its absence (Windows) must not break the import.
try:
    import fcntl
except ImportError:  # Windows
    fcntl = None  # type: ignore[assignment]

# ---------------------------------------------------------------------------
# Platform sentinel
# ---------------------------------------------------------------------------

_DARWIN: bool = sys.platform == "darwin"


def _full_flush_disabled() -> bool:
    """True when the caller opted out of the macOS full drive-cache flush.

    Set ``LILLI_FSYNC_MODE=fast`` to substitute a plain ``os.fsync`` for
    ``F_FULLFSYNC``. Intended for test and CI runs, where logical correctness —
    not power-loss durability — is under test: the per-commit ``F_FULLFSYNC``
    latency otherwise makes large property-test runs take hours. Leave unset in
    production so committed pages reach NAND.
    """
    return os.environ.get("LILLI_FSYNC_MODE") == "fast"


# ---------------------------------------------------------------------------
# Durable sync helpers
# ---------------------------------------------------------------------------


def durable_fsync(fd: int) -> None:
    """Flush fd to stable storage.

    On macOS, F_FULLFSYNC forces the drive controller's volatile DRAM cache
    to NAND. On Linux, os.fsync issues FLUSH to the storage device. Falls
    back to os.fsync if F_FULLFSYNC is unavailable (network FS, FUSE, VMs).
    Raises OSError on failure — callers must not silence this.
    """
    if _DARWIN and not _full_flush_disabled():
        try:
            ret = fcntl.fcntl(fd, fcntl.F_FULLFSYNC)
            if ret != 0:
                os.fsync(fd)
        except (AttributeError, OSError):
            os.fsync(fd)
    else:
        os.fsync(fd)


def durable_fdatasync(fd: int) -> None:
    """Flush fd data to stable storage, skipping metadata if supported.

    On macOS, F_FULLFSYNC is used (no fdatasync equivalent). On Linux,
    os.fdatasync skips non-critical metadata updates for performance.
    Falls back to os.fsync when fdatasync is unavailable (macOS, FUSE).
    Raises OSError on failure — callers must not silence this.
    """
    if _DARWIN and not _full_flush_disabled():
        try:
            ret = fcntl.fcntl(fd, fcntl.F_FULLFSYNC)
            if ret != 0:
                os.fsync(fd)
        except (AttributeError, OSError):
            os.fsync(fd)
    else:
        _fdatasync = getattr(os, "fdatasync", None)
        if _fdatasync is not None:
            _fdatasync(fd)
        else:
            os.fsync(fd)


def fsync_directory(path: str | Path) -> None:
    """Flush the directory entry for ``path`` to stable storage.

    Ensures that newly created files in the directory are durable after a
    rename/unlink sequence. The directory fd is opened read-only, fsync'd,
    then closed. Some platforms (macOS) silently succeed on directory fsync
    without flushing; callers must also fsync the file itself when durability
    of the file content is required.

    Raises OSError if the directory cannot be opened or flushed.

    No-op on Windows: os.open() on a directory raises PermissionError there,
    and the platform exposes no directory handle to fsync. Metadata durability
    for the rename itself is NTFS's business, so skipping is the only correct
    behaviour — raising would take the caller's whole checkpoint down with it.
    """
    if os.name == "nt":
        return
    dir_path = str(path)
    dir_fd = os.open(dir_path, os.O_RDONLY)
    try:
        os.fsync(dir_fd)
    finally:
        os.close(dir_fd)

File: crates/lillibrain/src/io.rs
//! Durable file-sync primitives.
//!
//! On macOS a plain `fsync` only pushes data to the drive controller; the
//! controller's volatile DRAM cache is flushed to stable NAND only by
//! `F_FULLFSYNC`. This module wraps that distinction so callers never invoke
//! `fsync` directly, and propagates every failure — an fsync error is never
//! swallowed.
//!
//! Setting the environment variable `LILLI_FSYNC_MODE=fast` substitutes a plain
//! data sync for the full drive-cache flush. This trades power-loss durability
//! for speed and is intended only for test and continuous-integration runs,
//! where the per-commit full-flush latency would otherwise dominate.

use std::fs::File;
use std::path::Path;

use crate::error::Result;

/// True when the caller opted out of the full drive-cache flush.
fn full_flush_disabled() -> bool {
    std::env::var("LILLI_FSYNC_MODE")
        .map(|v| v == "fast")
        .unwrap_or(false)
}

/// Flush a file to stable storage.
///
/// On macOS this issues `F_FULLFSYNC`, which forces the drive controller's
/// volatile cache to NAND, falling back to `sync_all` when the platform rejects
/// it (network filesystems, FUSE, virtual machines). On other platforms it
/// issues `sync_all`. When `LILLI_FSYNC_MODE=fast` is set it issues `sync_data`.
/// The error is always propagated.
pub fn durable_fsync(file: &File) -> Result<()> {
    if full_flush_disabled() {
        file.sync_data()?;
        return Ok(());
    }
    full_fsync(file)
}

#[cfg(target_os = "macos")]
fn full_fsync(file: &File) -> Result<()> {
    use std::os::unix::io::AsRawFd;
    let fd = file.as_raw_fd();
    // SAFETY: fd is a valid open file descriptor owned by `file` for the call.
    let ret = unsafe { libc::fcntl(fd, libc::F_FULLFSYNC) };
    if ret != 0 {
        // Full flush is unsupported on this volume; fall back to a normal sync.
        file.sync_all()?;
    }
    Ok(())
}

#[cfg(not(target_os = "macos"))]
fn full_fsync(file: &File) -> Result<()> {
    file.sync_all()?;
    Ok(())
}

/// Flush the directory entry for `path` to stable storage.
///
/// Ensures that a newly created, renamed, or unlinked file in the directory is
/// durable. The directory is opened read-only, synced, and closed. The error is
/// always propagated.
pub fn fsync_directory(path: &Path) -> Result<()> {
    // Windows: File::open on a directory needs FILE_FLAG_BACKUP_SEMANTICS and
    // returns ERROR_ACCESS_DENIED (os error 5) without it, which took the
    // journal create -- and with it every native-store open -- down on
    // SERVER-ALPHA. Directory-entry durability for the rename is NTFS's
    // business, so skipping is the correct behaviour, mirroring the Python
    // twin in lillibrain/io.py.
    #[cfg(windows)]
    {
        let _ = path;
        Ok(())
    }
    #[cfg(not(windows))]
    {
        let dir = File::open(path)?;
        dir.sync_all()?;
        Ok(())
    }
}

/// Flush the directory entry for `path`'s containing directory to stable
/// storage. An empty parent (bare relative path) resolves to the current
/// directory. The error is always propagated.
pub fn fsync_parent_directory(path: &Path) -> Result<()> {
    let parent = match path.parent() {
        Some(p) if !p.as_os_str().is_empty() => p,
        _ => Path::new("."),
    };
    fsync_directory(parent)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    #[test]
    fn io_durable_fsync_full_mode_returns_ok() {
        // Ensure the fast escape is off for this assertion.
        std::env::remove_var("LILLI_FSYNC_MODE");
        let mut f = tempfile::tempfile().unwrap();
        f.write_all(b"hello world").unwrap();
        durable_fsync(&f).unwrap();
    }

    #[test]
    fn io_durable_fsync_fast_mode_returns_ok() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("fast.bin");
        let mut f = File::create(&path).unwrap();
        f.write_all(b"fast path").unwrap();
        // Drive the fast path explicitly rather than mutating process env,
        // which would race other tests in the same binary.
        f.sync_data().unwrap();
        durable_fsync(&f).unwrap();
    }

    #[test]
    fn io_fsync_directory_returns_ok() {
        let dir = tempfile::tempdir().unwrap();
        fsync_directory(dir.path()).unwrap();
    }
}
